File: go/src/RumbleRacing/core/asset/RLst.go
package asset

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
	"strings"
)

type RLst struct {
	FileName string
	Count    uint32
	Entries  []ResourceEntry
}

func (r *RLst) GetType() string {
	return "RLst"
}

type ResourceEntry struct {
	TypeTag       string
	ResourceIndex uint32
	ResourceName  string
}

func ParseRLst(data []byte, fileName string) (*RLst, error) {
	r := bytes.NewReader(data)

	var count uint32
	if err := binary.Read(r, binary.LittleEndian, &count); err != nil {
		return nil, fmt.Errorf("failed to read count: %w", err)
	}

	entries := make([]ResourceEntry, 0, count)
	for i := uint32(0); i < count; i++ {
		var tag [4]byte
		if _, err := io.ReadFull(r, tag[:]); err != nil {
			return nil, fmt.Errorf("failed to read tag (entry %d): %w", i, err)
		}

		// Reverse bytes
		for j := 0; j < 2; j++ {
			tag[j], tag[3-j] = tag[3-j], tag[j]
		}

		var index uint32
		if err := binary.Read(r, binary.LittleEndian, &index); err != nil {
			return nil, fmt.Errorf("failed to read index (entry %d): %w", i, err)
		}

		var nameBytes [24]byte
		if _, err := io.ReadFull(r, nameBytes[:]); err != nil {
			return nil, fmt.Errorf("failed to read name (entry %d): %w", i, err)
		}

		name := string(nameBytes[:])
		name = strings.TrimRight(name, "\x00") // strip null padding

		entry := ResourceEntry{
			TypeTag:       string(tag[:]),
			ResourceIndex: index,
			ResourceName:  name,
		}
		entries = append(entries, entry)
	}

	return &RLst{
		Count:    count,
		Entries:  entries,
		FileName: fileName,
	}, nil
}

File: src/RumbleRacing/asset/RLst.ts
import ArrayBufferSlice from "../../ArrayBufferSlice";
import { readString } from "../../util";
import { readFourCC } from "../helpers/fourCC";

export interface ResourceEntry {
  typeTag: string;
  resourceIndex: number;
  resourceName: string;
}

export interface RLst {
  fileName: string;
  count: number;
  entries: ResourceEntry[];
}

export function parseRLst(data: ArrayBufferSlice, fileName: string): RLst {
  const view = data.createDataView();
  let pos = 0;

  const count = view.getUint32(pos, true);
  pos += 4;

  const entries: ResourceEntry[] = [];

  for (let i = 0; i < count; i++) {
    const typeTag = readFourCC(data, pos);
    pos += 4;

    const index = view.getUint32(pos, true);
    pos += 4;

    const name = readString(data, pos, 24, true);
    pos += 24;

    entries.push({ typeTag, resourceIndex: index, resourceName: name });
  }

  return { fileName, count, entries };
}
